controllers/topology/desired_state.go

```diff
@@ -302,6 +340,55 @@ func computeMachineDeployment(_ context.Context, s *scope.Scope, machineDeployme
 	return desiredMachineDeployment, nil
 }
 
+// computeMachineDeploymentVersion returns the desired machine deployment version
+// If creating a new machine deployment, returns the version defined in topology.
+// If a machine deployment already exists:
+// - Returns the version from the current machine deplyment if any of the following are true:
+//    - If a control plane already exists and is in between an upgrade.
+//    - If a control plane already exists and is ready to upgrade.
+//    - Any of the machine deployments are in between an upgrade.
+func computeMachineDeploymentVersion(s *scope.Scope, desiredControlPlaneState *scope.ControlPlaneState, currentMDState *scope.MachineDeploymentState) (string, error) {
+	topologyVersion := s.Blueprint.Topology.Version
+	// if creating a new machine deployment, use version defined in topology
+	if currentMDState == nil || currentMDState.Object == nil {
+		return topologyVersion, nil
+	}
+
+	currentVersion := *currentMDState.Object.Spec.Template.Spec.Version
+
+	// do not allow version downgrades
+	if err := isDowngrade(currentVersion, topologyVersion); err != nil {
+		return currentVersion, errors.Wrap(err, "failed version downgrade check")
+	}
+
+	var cpUpgrading bool
+	var err error
+	// If a control plane already exists, check to see if it is upgrading or about to upgrade
+	if s.Current.ControlPlane != nil && s.Current.ControlPlane.Object != nil {
+		cpUpgrading, err = contract.ControlPlane().Upgrading(s.Current.ControlPlane.Object)
+		if err != nil {
+			return currentVersion, errors.Wrap(err, "failed to check if control plane is upgrading")
+		}
+		if !cpUpgrading {
+			// control plane is not in the middle of an upgrade, checking if it is about to upgrade
+			currentCPVersion, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
+			if err != nil {
+				return currentVersion, errors.Wrap(err, "failed to get version of current control plane")
+			}
+			computedCPVersion, err := contract.ControlPlane().Version().Get(desiredControlPlaneState.Object)
+			if err != nil {
+				return currentVersion, errors.Wrap(err, "failed to get version of desired control plane")
+			}
+			cpUpgrading = *currentCPVersion != *computedCPVersion
+		}
+	}
+
+	if cpUpgrading || s.Current.MachineDeployments.Upgrading() {
+		return currentVersion, nil
+	}
+	return topologyVersion, nil
```

Comment:
nit Let's try to rewrite this avoiding the nested if, and adding verbose comments for future maintainability